XO.py

```diff
@@ -0,0 +1,157 @@
+from random import shuffle
+
+# size of board
+N = 3
+
+# Utility functions
+
+
+def print_board(board):
+    print()
+    print('Board:')
+    for line_ in board:
+        for item in line_:
+            print(item, end='')
+        print()
+    print()
+
+
+def check_input_validity(input_type, curr_player):
+    not_valid = True
+    while not_valid:
+        print('{}, please insert the {} number'.format(curr_player, input_type))
+        try:
+            input_value = int(input())
+        except:
+            print('{}, please insert a valid {} number'.format(curr_player, input_type))
+            continue
+        if input_value < 0 or input_value >= N:
+            print('{}, please insert a valid {} number'.format(curr_player, input_type))
+            continue
+        not_valid = False
+    return input_value
+
+
+def check_line(board, line, curr_player, curr_player_symbol, is_win):
+    for sym in board[line]:
+        if sym != curr_player_symbol:
+            is_win = False
+            break
+    if is_win:
+        return declare_win(curr_player)
+
+
+def check_col(board, col, curr_player, curr_player_symbol, is_win):
+    for line in board:
+        if line[col] != curr_player_symbol:
+            is_win = False
+            break
+    if is_win:
+        return declare_win(curr_player)
+
+
+def check_main_diagonal(board, line, col, curr_player, curr_player_symbol, is_win):
+    if line == col:
+        for i in range(N):
+            if board[i][i] != curr_player_symbol:
+                is_win = False
+                break
+    else:
+        is_win = False
+
+    if is_win:
+        return declare_win(curr_player)
+
+
+def check_opposite_diagonal(board, line, col, curr_player, curr_player_symbol, is_win):
+    if line + col == N - 1:
+        j = 2

```

Comment:
Why 2? Shouldn't this be related to `N` somehow?

Reply:
Right... nasty mistake. Fixed.